Update konstruktoid/ansible-role-hardening for cisagov's needs

bra1ncramp opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

Update the ansible-role-hardening-2 for cisagov's needs.

## 💭 Motivation and context ##

Our current ansible-role-hardening was forked from [konstruktoid/ansible-role-hardening](https://github.com/konstruktoid/ansible-role-hardening) a few years ago. Since then, significant changes and improvements have been made in the original code. 

When building out a new Ubuntu AMI, we found that our existing hardening role was not sufficiently hardening the system, so we forked the current code from [konstruktoid/ansible-role-hardening](https://github.com/konstruktoid/ansible-role-hardening) and edited it to meet our use cases.

Resolves cisagov/freeipa-server-packer#117.

## 🧪 Testing ##

We built out an Ubuntu server and applied this role to it. We were able to successfully access the server and pass Nessus compliance scans.

We have also [tested on a Debian server](https://github.com/cisagov/openvpn-packer/pull/93) as well as [a Fedora server](https://github.com/cisagov/freeipa-server-packer/pull/113).


## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.

bra1ncramp committed:
Change umask from 077 to 027

bra1ncramp committed:
Add tasks for securing SSH config options

bra1ncramp committed:
Ensure that umask lines appear in all files checked by Nessus

bra1ncramp committed:
Ensure that TMOUT is set everywhere the Nessus scan checks

bra1ncramp committed:
Ensure logrotate applies correct permissions

bra1ncramp committed:
Configure /etc/login.defs with reasonable values

bra1ncramp committed:
Move umask tasks to run nearly last

bra1ncramp committed:
Configure AWS preferred ntp servers

bra1ncramp committed:
Modify ssh session settings

bra1ncramp committed:
Add tasks to set ssh ciphers, MACs, and kex algs for RedHat

bra1ncramp committed:
Only apply the timesyncd config changes if the config file exists

bra1ncramp committed:
Removed unwanted tasks from running

bra1ncramp committed:
Remove uneccesary ssh config

Protocol 1 has been disabeled and Protocol 2 is assumed. It no longer needs to be explicitly declared.

bra1ncramp committed:
Clean up crypto policies

bra1ncramp committed:
Implement Upstream changes to crypto-policy config

bra1ncramp committed:
Remove firewall tasks

We'll implement these separately

bra1ncramp committed:
Disable ufw setup

We will do this ourselves

bra1ncramp committed:
Remove restrictions to sshd_admin_net

bra1ncramp committed:
Add required packages

bra1ncramp committed:
Adding packages to prevent null values

bra1ncramp committed:
Disable AIDE install temporarily

bra1ncramp committed:
Add package cracklib-runtime for handler cleanup

bra1ncramp committed:
Add required packages for RedHat

bra1ncramp committed:
Enable auditd tasks

Post checks fail if auditd isn't set in grub.cfg

bra1ncramp committed:
Add audit packages

bra1ncramp committed:
Set sshd client alive settings per CIS guidelines

bra1ncramp committed:
Ensure password reuse is limited

According to CIS benchmarks, we need to have the `remember=5` on the `pam_unix.so` line

bra1ncramp committed:
Remove audispd from audit tool verification

bra1ncramp committed:
Enable AIDE install

bra1ncramp committed:
Move ssh and pam rules to after aide config

bra1ncramp committed:
Limit aide checksums

bra1ncramp committed:
Call update-grub after updating or creating a grub configuration

This is backported from upstream.
https://github.com/cisagov/ansible-role-hardening-2/commit/8c6808ffa7a2f97ee350ff583dbd21f5dfd6928e

bra1ncramp committed:
Create aide defaults file

bra1ncramp committed:
add aide config validation

bra1ncramp committed:
Dont limit, configure

bra1ncramp committed:
Reorder task for consistancy with upstream

bra1ncramp committed:
Handle absent crontab

bra1ncramp committed:
Handle missing cron directories

bra1ncramp committed:
Add template variables

bra1ncramp committed:
Update documentation

bra1ncramp committed:
Remove examples from documentation

bra1ncramp committed:
Update hardening.rules.j2

Remove URL

bra1ncramp committed:
Improvements restrict compilers

Use `dpkg-statoverride` and `DNF post-transaction-actions Plugin` to ensure restrictive permissions on binaries during package installation and upgrades.

More details:
- https://man7.org/linux/man-pages/man1/dpkg-statoverride.1.html
- https://dnf-plugins-core.readthedocs.io/en/latest/post-transaction-actions.html

bra1ncramp committed:
Fix implicit octal value

bra1ncramp committed:
Fix privilege escalation of task

bra1ncramp committed:
update suid list from @GTFOBins

jsf9k committed:
Comment out haveged handler notification

The haveged package was commented out in tasks/packages.yml by
@bra1ncramp in commit bbffe17e, so we shouldn't call the handler.

jsf9k committed:
Disable the systemd-timesyncd functionality

We (cisagov) don't want this functionality since we configure chrony
on all our instances and point it to the AWS time source.

jsf9k commented:
- [I verified that using these changes results in an OpenVPN AMI that functions as expected](https://github.com/cisagov/openvpn-packer/pull/93).
- [These changes uninstall `freeipa-server` and dependencies from the FreeIPA AMI](https://github.com/cisagov/freeipa-server-packer/pull/113) for some reason I've yet to skry.
- [I verified that using these changes results in a FreeIPA AMI that functions as expected _if I first reinstall `freeipa-server`_](https://github.com/cisagov/freeipa-server-packer/pull/113).

jsf9k committed:
Remove from blocklist packages required for freeipa-server

Uninstalling these packages will also uninstall freeipa-client and
freeipa-server on Fedora.

jsf9k committed:
Remove snapd package from Ubuntu

The package was not desired but still listed because someone did not
know how to create an empty list (versus a null value) in YAML.

jsf9k committed:
Update README for recent changes to defaults/main/packages.yml

jsf9k reviewed (changes requested):
In general, you should comment your deviations from [konstructoid/ansible-role-hardening](https://github.com/konstructoid/ansible-role-hardening) stating why they were necessary.

jsf9k commented on tasks/auditd.yml:
```diff
@@ -10,6 +10,8 @@
     owner: root
     group: root
   when: ansible_os_family == "Debian"
+  notify:
+    - Update GRUB
```
Why is this necessary?

jsf9k replied:
I see that this was done as part of commit 452e5c9aecf680b6436ccca439f857173c691cdf, where you are updating `grub` anytime the `grub` configuration is modified.  Can you add a comment to that effect?

jsf9k replied:
Taken care of in commit 6e19020.

jsf9k commented on tasks/compilers.yml:
```diff
@@ -20,7 +20,23 @@
+  tags:
+    - compilers
+
+- name: Ensure restrict compilers access via dpkg-statoverride
+  become: true
+  ansible.builtin.command: dpkg-statoverride --update --force-all --add root root 0750 "{{ item.path }}"
+  register: dpkg_statoverride
+  changed_when: dpkg_statoverride.rc != 0
+  loop: "{{ compiler.files }}"
+  when: ansible_os_family == 'Debian'
+  tags:
+    - compilers
+
+- name: Ensure restrict compilers access via DNF post-transaction-actions Plugin
+  ansible.builtin.include_tasks: compilers_dnf_post_transaction_actions_plugin.yml
+  loop: "{{ compiler.files | map(attribute='path') | unique }}"
+  when: ansible_os_family == 'RedHat'
```
Why was this added?

jsf9k replied:
It looks like this was added in commit c8bb4769a805b8e9cc5dbc817e845bdee63c591b to further lock down compilers.  Can you add a comment explaining why that is a good idea?

bra1ncramp replied:
I have continued to backport commits from upstream. See https://github.com/cisagov/ansible-role-hardening-2/commit/0ccbf12f1d214c61abd77db5baf5ede724c3476e

jsf9k replied:
OK, but for future reference the way to backport commits from upstream is to either:
- `git rebase` on top of the upstream
- `git cherry-pick` from upstream

You shouldn't recreate the commits yourself.

bra1ncramp replied:
I'm always learning

jsf9k commented on tasks/ipv6.yml:
```diff
@@ -10,6 +10,8 @@
     owner: root
     group: root
   when: ansible_os_family == "Debian"
+  notify:
+    - Update GRUB
```
Why was this added?

jsf9k replied:
I see that this was done as part of commit 452e5c9aecf680b6436ccca439f857173c691cdf, where you are updating `grub` anytime the `grub` configuration is modified.  Can you add a comment to that effect?

jsf9k replied:
Taken care of in commit 6e19020.

jsf9k commented on tasks/timesyncd.yml:
```diff
@@ -4,21 +4,29 @@
   ansible.builtin.package:
     name: systemd-timesyncd
     state: present
+  when: ansible_os_family == "Debian"  
```
Why was this added?

jsf9k replied:
Why is this change part of commit 573b13ff3951526b98100e4d7c3eccd0379c5907?  The commit message for that particular commit doesn't mention anything that explains this change.

bra1ncramp replied:
Per this commit, timesyncd does not exist on Fedora
https://github.com/konstruktoid/ansible-role-hardening/commit/e238a9d2be1a0ac8e4b9103e58456e2449d7129f

jsf9k replied:
[The commit you referenced](https://github.com/konstruktoid/ansible-role-hardening/commit/e238a9d2be1a0ac8e4b9103e58456e2449d7129f) does not contain this change.  That commit also states that the file _doesn't exist by default_ on Fedora 35; it doesn't say that the package doesn't exist at all.  In any event, [Fedora 35 is EOL as of almost a year ago](https://docs.fedoraproject.org/en-US/releases/eol/).

Also, please see [this comment](https://github.com/cisagov/ansible-role-hardening-2/pull/1#discussion_r1388217368), as it applies here as well.

jsf9k replied:
I think you should remove this line, in the interest of staying as true to the upstream as possible and therefore making the pulling in of upstream commits in the future as smooth as possible.

bra1ncramp replied:
Removed line in [b353fe0](https://github.com/cisagov/ansible-role-hardening-2/pull/1/commits/b353fe083b358cca07f3cc643adc6ce671121068)

jsf9k replied:
Thank you!

jsf9k commented on tasks/timesyncd.yml:
```diff
@@ -4,21 +4,29 @@
     - CIS-UBUNTU2004-2.1.1.1
     - CIS-UBUNTU2004-2.1.1.2
 
+- name: Stat systemd timesyncd
+  become: true
+  ansible.builtin.stat:
+    path: /etc/systemd/timesyncd.conf
+  register: timesyncd   
+
```
Why did this need to be added?

jsf9k replied:
I see this was added in 573b13ff3951526b98100e4d7c3eccd0379c5907 so that the modifications to the `timesyncd` configuration file are only made if the configuration file actually exists.  Can you add a comment to that effect?

jsf9k replied:
Taken care of in commit 76e0283.

jsf9k commented on templates/etc/audit/rules.d/hardening.rules.j2:
```diff
@@ -1,5 +1,4 @@
 # {{ ansible_managed }}
-# https://github.com/konstruktoid/ansible-role-hardening
```
Why was this removed?

bra1ncramp replied:
This was backported from upstream commit: https://github.com/cisagov/ansible-role-hardening-2/commit/b3571d307ab670304eba4e9cf3886e1bda99a3db

jsf9k replied:
OK, but please see [this comment](https://github.com/cisagov/ansible-role-hardening-2/pull/1#discussion_r1388217368), as it applies here as well.

jsf9k reviewed (changes requested):
I called out several places where I think a comment is really necessary to explain why lines were added or removed.

jsf9k commented:
> * [I verified that using these changes results in an OpenVPN AMI that functions as expected](https://github.com/cisagov/openvpn-packer/pull/93).
> * [These changes uninstall `freeipa-server` and dependencies from the FreeIPA AMI](https://github.com/cisagov/freeipa-server-packer/pull/113) for some reason I've yet to skry.
> * [I verified that using these changes results in a FreeIPA AMI that functions as expected _if I first reinstall `freeipa-server`_](https://github.com/cisagov/freeipa-server-packer/pull/113).

Update:
* [I verified that using these changes results in an OpenVPN AMI that functions as expected](https://github.com/cisagov/openvpn-packer/pull/93).
* ~[These changes uninstall `freeipa-server` and dependencies from the FreeIPA AMI](https://github.com/cisagov/freeipa-server-packer/pull/113) for some reason I've yet to skry.~
* [I verified that using these changes results in a FreeIPA AMI that functions as expected ~_if I first reinstall `freeipa-server`_~](https://github.com/cisagov/freeipa-server-packer/pull/113).

jsf9k committed:
Add a comment stating why a notification was added

jsf9k committed:
Add a comment explaining the addition of a stat module

It is being added because we only want to modify the timesyncd
configuration if it actually exists.

jsf9k commented:
@bra1ncramp - I resolved the conversations I could take care of by myself.  You'll have to take care of the rest before we can merge this.

bra1ncramp committed:
Remove when Debian

This conditional is unnecessary and devient from upstream

jsf9k reviewed (approved):
Thank you @bra1ncramp!  Please make sure to read [this comment](https://github.com/cisagov/ansible-role-hardening-2/pull/1/files#r1388217368) for future reference regarding pulling in upstream changes.

I think we can merge once we get an approving review from @dav3r.

dav3r commented on tasks/aide.yml:
```diff
@@ -70,14 +70,28 @@
     - /usr/sbin/autrace
     - /usr/sbin/augenrules
-    - /usr/sbin/audispd
+    # - /usr/sbin/audispd # This is not required for CIS Benchmarks, so we won't be installing audispd
```
Remove unneeded capitalization in a comment:
```suggestion
    # - /usr/sbin/audispd # This is not required for CIS benchmarks, so we won't be installing audispd
```

dav3r commented on README.md:
```diff
@@ -356,53 +326,55 @@ packages_blocklist:
-  - apparmor-utils
-  - apt-show-versions
-  - audispd-plugins
+  # - acct
+  # - apparmor-profiles
+  # - apparmor-utils
+  # - apt-show-versions
+  # - audispd-plugins
```
Why are these packages commented out?  Note: this goes for all of the commented-out packages in this file. There should be some explanation of why.

bra1ncramp replied:
See 94f7b7442d3c7ad4e858678561ac7fb081005155

dav3r commented on defaults/main/packages.yml:
```diff
@@ -22,50 +25,52 @@ packages_blocklist:
-  - apparmor-utils
-  - apt-show-versions
-  - audispd-plugins
+  # - acct
+  # - apparmor-profiles
+  # - apparmor-utils
+  # - apt-show-versions
+  # - audispd-plugins
```
Why are these packages commented out?  Note: this goes for all of the commented-out packages in this file.  There should be some explanation of why.

bra1ncramp replied:
see 94f7b7442d3c7ad4e858678561ac7fb081005155

dav3r commented on tasks/main.yml:
```diff
@@ -2,19 +2,19 @@
-- name: Configure systemd resolved
-  ansible.builtin.include_tasks: resolvedconf.yml
+# - name: Configure systemd resolved
+#   ansible.builtin.include_tasks: resolvedconf.yml
```
The reason for each set of commented-out tasks should be explained via a comment.

bra1ncramp replied:
I would ask for Dev to assist with this, then - as I am following this:
https://github.com/cisagov/ansible-role-hardening/blob/master/tasks/main.yml

dav3r replied:
@bra1ncramp I think you can just put something generic for each commented-out section like:
```
# These tasks are not currently needed for our use case
```

bra1ncramp replied:
see cb81d61417d3a3a7e19bce5096868e9a88b645db

dav3r reviewed (commented):
I have some questions before I approve.  My apologies if I missed any previous discussions about this stuff.

bra1ncramp committed:
Remove unneeded capitalization in a comment

Co-authored-by: dav3r <daver@geekpad.com>

bra1ncramp committed:
Add clarifying comment

jsf9k committed:
If /home is a mount, ensure it is mounted with the nodev option

Resolves cisagov/freeipa-server-packer#117.

bra1ncramp committed:
Add comments on disabled tasks

dav3r reviewed (approved):
Thanks for sticking with this one @bra1ncramp - everything LGTM now, nice work! 👍

jsf9k commented:
@bra1ncramp - Please don't merge this PR yet.  I am still testing the change from commit dd19901.

jsf9k commented:
> @bra1ncramp - Please don't merge this PR yet. I am still testing the change from commit [dd19901](https://github.com/cisagov/ansible-role-hardening-2/commit/dd1990132a3092455039a78efbe49af172257453).

OK, I think this one is good to merge now.  It took several attempts to get the `nodev` jazz working, but commit 380b2b3 does the trick.

Do you agree that it's ready to merge, @bra1ncramp?  If so then either merge or let me know and I'll click the button.

konstruktoid commented:
Hi and sorry for commenting like this.

I'm planning to release 2.0.0 due to breaking changes (see eg https://github.com/konstruktoid/ansible-role-hardening/blob/8f537cdba94604be52fe33c726ae1c9aa61a06a3/defaults/main/sshd.yml) in a week or so, and since you're the most active fork, I was wondering if there any changes you would like implemented?

jsf9k commented:
> Hi and sorry for commenting like this.
> 
> I'm planning to release 2.0.0 due to breaking changes (see eg https://github.com/konstruktoid/ansible-role-hardening/blob/8f537cdba94604be52fe33c726ae1c9aa61a06a3/defaults/main/sshd.yml) in a week or so, and since you're the most active fork, I was wondering if there any changes you would like implemented?

Hi @konstruktoid!  I don't have any changes right now, but I may in the future.  I want to move away from this fork and instead use your role directly, which may entail adding a few role variables.  (See cisagov/freeipa-server-packer#118 and cisagov/openvpn-packer#96.)

konstruktoid commented:
That's great to hear @jsf9k 

I'll take a look at the variables, and you're of course welcome to submit any PRs as well.

konstruktoid commented:
It took a bit longer than a week...
v2.0.0-rc.1 released: https://github.com/konstruktoid/ansible-role-hardening/releases/tag/v2.0.0-rc.1
